File: src/day3b.rs
fn distance(p1:&(i32, i32), p2:&(i32, i32)) -> i32 {
    return (p1.0 - p2.0).abs() + (p1.1 - p2.1).abs();
}

fn convert_to_coordinates<'a, T: Iterator<Item=&'a str>>(moves: T) -> Vec<(i32, i32)> {
    let start = (0, 0);
    let mut res = Vec::new();
    res.push(start);
    for m in moves {
        let new_pos = {
            let current_pos = res.last().unwrap();
            let direction = &m[..1];
            let distance = m[1..].parse::<i32>().unwrap()
                * if ["R", "U"].contains(&direction) {1} else {-1};
            (
                current_pos.0 + if ["L", "R"].contains(&direction) {distance} else {0},
                current_pos.1 + if ["U", "D"].contains(&direction) {distance} else {0})
        };
        res.push(new_pos);
    }

    return res
}

fn main() {
    use std::io::{self, BufRead};
    use std::cmp;

    let stdin = io::stdin();

    let wires = (0..2).map(
        |_| convert_to_coordinates(
            stdin.lock().lines().next().unwrap().unwrap().split(",")))
        .collect::<Vec<_>>();

    let sections = wires.iter().map(|wire| wire.iter().zip(wire.iter().skip(1))
                                    .collect::<Vec<_>>())
        .collect::<Vec<_>>();

    let dist = sections.iter().map(|section| section.iter().scan(
            0, |d, &s| {
                let d1 = *d;
                *d = *d + distance(s.0, s.1);
                Some(d1)
            }).collect::<Vec<_>>())
        .collect::<Vec<_>>();

    let res = sections[0].iter().zip(dist[0].iter()).filter_map(
        |(section1, dist1)| sections[1].iter().zip(dist[1].iter()).filter_map(
            |(section2, dist2)| {
                let (x1, x2, x, y1, y2, y) = if (section1.0).0 == (section1.1).0 {
                    ((section2.0).0, (section2.1).0, (section1.0).0,
                     (section1.0).1, (section1.1).1, (section2.0).1)
                }
                else {
                    ((section1.0).0, (section1.1).0, (section2.0).0,
                     (section2.0).1, (section2.1).1, (section1.0).1)
                };

                let (x1, x2) = (cmp::min(x1, x2), cmp::max(x1, x2));
                let (y1, y2) = (cmp::min(y1, y2), cmp::max(y1, y2));

                if x1 < x && x < x2 && y1 < y && y < y2 {
                    Some(dist1 + distance(section1.0, &(x, y))
                        + dist2 + distance(section2.0, &(x, y)))
                }
                else {
                    None
                }
            }).min()).min().unwrap();

    println!("{}", res);
}

File: src/day3b.py
def dist(p1, p2):
    return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])

def convert_to_coordinates(moves, start=(0, 0)):
    res = [start]
    for move in moves:
        current_pos = res[-1]
        direction = move[0]
        dist = int(move[1:]) * (1 if direction in ['R', 'U'] else -1)
        res.append(
                (
                    current_pos[0] + (dist if direction in ['L', 'R'] else 0),
                    current_pos[1] + (dist if direction in ['U', 'D'] else 0),
                )
            )
    return res

wires = [convert_to_coordinates(input().split(',')) for _ in range(2)]

min_dist = None

dist1 = 0

for section1 in zip(wires[0], wires[0][1:]):
    dist2 = 0
    for section2 in zip(wires[1], wires[1][1:]):
        if section1[0][0] == section1[1][0]: # horizontal section
            x, y1 = section1[0]
            _, y2 = section1[1]

            x1, y = section2[0]
            x2, _ = section2[1]
        else: #vertical section
            x1, y = section1[0]
            x2, _ = section1[1]

            x, y1 = section2[0]
            _, y2 = section2[1]

        x1, x2 = min(x1, x2), max(x1, x2)
        y1, y2 = min(y1, y2), max(y1, y2)

        if x1 < x < x2 and y1 < y < y2:
            dist_inter = dist1 + dist(section1[0], (x, y))\
                       + dist2 + dist(section2[0], (x, y))

            try:
                min_dist = min(min_dist, dist_inter)
            except TypeError:
                min_dist = dist_inter

        dist2 += dist(*section2)
    dist1 += dist(*section1)

print(min_dist)

